lib/stages/RfiFrameDrop.cpp

```diff
@@ -123,78 +128,90 @@ void RfiFrameDrop::main_thread() {
         DEBUG2("Frames are synced. Vis frame: {}; SK frame: {}, diff {}", vis_seq, sk_seq,
                vis_seq - sk_seq);
 
-        // Point the shared pointer to the newest updates
-        thresholds = std::atomic_load(&_thresholds);
-
         // Calculate the scaling to turn kurtosis value into sigma
         size_t num_inputs = num_elements - metadata_vis->rfi_num_bad_inputs;
         float sigma_scale = sqrt((num_inputs * (sk_step - 1) * (sk_step + 2) * (sk_step + 3))
                                  / (4.0 * sk_step * sk_step));
 
-        for (size_t ii = 0; ii < num_sub_frames; ii++) {
+        {
+            // Lock update mutex to not allow updates being processed during this critical section
+            std::lock_guard<std::mutex> lock(update_mutex);
```

Comment:
The lifetime of this lock seems to be longer than required. I think you only need to hold it long enough to get a copy of the dataset_id and the thresholds and num_sk values.